src/Stratis.Features.FederatedPeg/FederationGatewayFeature.cs

```diff
@@ -306,6 +306,9 @@ public static IFullNodeBuilder AddFederationGateway(this IFullNodeBuilder fullNo
                         // Set up events.
                         services.AddSingleton<TransactionObserver>();
                         services.AddSingleton<BlockObserver>();
+
+                        // After setting up, invoke any additional options which can replace services as required.
+                        options?.Invoke(new FederatedPegOptions(services, fullNodeBuilder.NodeSettings));
```

Comment:
Not sure I like this trick. The general pattern should be for things like `options` (settings) to be used earlier (at the start of the method). Here it will be delayed. There are other mechanism for replacing services.

Reply:
@quantumagi I don't quite follow - I'm very happy to move them up to the start of the method but could you elaborate why?

Reply:
@codingupastorm , i may be wrong but I believe that the code is currently in the correct place to replace services (as per the accompanying comment). However, if you move it up to where one would typically expect to see the settings overrides, then it probably won't be. That is why I am suggesting using some of the expected mechanisms to override services when and if that is required - i.e. _not doing it via options_.

Reply:
Funnily enough that method doesn't actually replace services, so I'll fix the comment.

With that in mind does it make sense to leave the code though? It injects services based on the parameters, it's a bit gross but I'm not sure where else to put this logic (except maybe in the daemon but I'm not a fan of that)

Reply:
Perhaps once could add an `InjectService` method, loosely based on the `ReplaceService` code?

Reply:
I'm confused, how is InjectService different to using AddSingleton?

Reply:
I imagine the only difference would be in how its called - i.e. on the builder with `.InjectService(someService)` which would call `AddSingleton` under the covers. However as mentioned above the best option may be to have an overridable setting that determines which service to use.